### PR TITLE
feat: add method `receiver` edge

### PR DESCRIPTION
**Pull request by Frank-III**
finish on top of: #806 

one thing I need to dive into would be that we wrap a lot of things in a `Rc`, is it because of how trustfall engine works, the examples there seem to do the same, but I think we don't need this for the `Receiver`?

### CODE REVIEW

File: src/adapter/edges.rs [hunk 1]
```diff
@@ -261,6 +262,35 @@ pub(super) fn resolve_function_like_edge<'a, V: AsVertex<Vertex<'a>> + 'a>(
...
+) -> ContextOutcomeIterator<'a, V, VertexIterator<'a, Vertex<'a>>> {
+    match edge_name {
+        "receiver" => resolve_neighbors_with(contexts, move |vertex| {
+            let origin = vertex.origin;
+            let method = vertex.as_function().expect("vertex was not a Function");
+
+            // Check if the first parameter is a self receiver
+            let receiver = method.sig.inputs.first().and_then(|(name, ty)| {
+                if name.starts_with("self") {
+                    Some(MethodSelfReceiver::new(ty))
```
Comment (obi1kenobi) [inline src/adapter/edges.rs]: Can you add a test case where a method's first argument is named e.g. `selfless` and see how that behaves? Because that shouldn't be a receiver, I think, and I suspect this will treat it as one.

File: src/adapter/edges.rs [hunk 2]
```diff
@@ -261,6 +262,35 @@ pub(super) fn resolve_function_like_edge<'a, V: AsVertex<Vertex<'a>> + 'a>(
...
+                    None
+                }
+            });
+
+            match receiver {
+                Some(receiver) => Box::new(std::iter::once(
+                    origin.make_method_receiver_vertex(receiver),
+                )),
+                None => Box::new(std::iter::empty()),
+            }
```
Comment (obi1kenobi) [inline src/adapter/edges.rs]: There's a cute trick you can do here: `Option<T>` has [`.into_iter()`](https://doc.rust-lang.org/std/option/enum.Option.html#method.into_iter) so you can do:
```suggestion
            Box::new(receiver.into_iter().map(|r| origin.make_method_receiver_vertex(r)))
```

File: src/adapter/method_self_receiver.rs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        matches!(
+            self.0,
+            Type::BorrowedRef {
+                is_mutable: true,
+                ..
+            }
+        )
+    }
+
+    pub(super) fn kind(&self) -> String {
```
Comment (obi1kenobi) [inline src/adapter/method_self_receiver.rs]: Let's have this and its helper function both return `Cow<'_, str>` instead, so that we don't have to always copy and reallocate strings when the exact string can be borrowed from `self.0`?

File: src/adapter/vertex.rs [hunk 4]
```diff
@@ -59,6 +59,9 @@ pub enum VertexKind<'a> {
     #[non_exhaustive]
     FunctionAbi(&'a Abi),
 
+    #[non_exhaustive]
+    MethodReceiver(MethodSelfReceiver<'a>),
```
Comment (obi1kenobi) [inline src/adapter/vertex.rs]: nit: The naming is pretty inconsistent from place to place.
- In the schema, the type is called `Receiver`.
- The variant is `MethodReceiver`.
- The internal type is `MethodSelfReceiver`.

I could possibly be convinced to have two different names instead of just `Receiver` everywhere. But three names for the same thing is definitely not what we want.

A priori, I still feel that `Receiver` is probably best as a one and only name, because while it's a bit rarely used in Rust, it's unambiguous. For example, a Google search for "receiver rust" returns info on the channel receiver type (obviously not a match) and [method receivers](https://google.github.io/comprehensive-rust/methods-and-traits/methods.html). But if you feel strongly about having different names in the schema and the code, let me know your thoughts — I could be convinced otherwise.

File: src/adapter/properties.rs [hunk 5]
```diff
@@ -306,6 +306,31 @@ pub(super) fn resolve_function_property<'a, V: AsVertex<Vertex<'a>> + 'a>(
     }
 }
 
+pub(super) fn resolve_method_receiver_property<'a, V: AsVertex<Vertex<'a>> + 'a>(
+    contexts: ContextIterator<'a, V>,
+    property_name: &str,
+) -> ContextOutcomeIterator<'a, V, FieldValue> {
+    match property_name {
+        "by_value" => resolve_property_with(contexts, |vertex| {
+            let receiver = vertex.as_method_receiver().expect("not a MethodReceiver");
```
Comment (obi1kenobi) [inline src/adapter/properties.rs]: This is an example of why uniform naming matters: `MethodReceiver` here is referencing the enum variant, but what it *actually* should be referencing is a type in the Trustfall schema. The issue is that we expected a different Trustfall schema type, and it just so happens the vertex variant type didn't match either.

The distinction is important: if we expected an enum but got a struct vertex, those are both `VertexKind::Item` or `VertexKind::PositionedItem` variants but we'll still panic about it regardless.

**Review by obi1kenobi — COMMENTED** [review]
Really nice work! This is 99% ready to merge, with just a handful of nitpicks to resolve first.

> one thing I need to dive into would be that we wrap a lot of things in a Rc, is it because of how trustfall engine works, the examples there seem to do the same, but I think we don't need this for the Receiver?

Trustfall will clone vertices often, so if a variant is expensive to clone then we wrap it in `Rc`. The new variant you added is cheap to clone so no `Rc` is needed.

File: src/adapter/edges.rs (hunk 1 above)
Comment (Frank-III) [reply src/adapter/edges.rs]: you are right!

File: src/adapter/vertex.rs (hunk 4 above)
Comment (Frank-III) [reply src/adapter/vertex.rs]: this was a mistake! I actually have the same thought, would change it to be the same as the schema

**Comment by Frank-III** [comment]
Thank you so much for the review! would fix it

File: test_crates/method_self_receivers/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+    pub fn by_pinned_mut_ref(self: std::pin::Pin<&mut Self>) {}
+
+    pub fn by_ref_pinned_mut_ref(self: &std::pin::Pin<&'_ mut Self>) {}
+
+    pub fn by_ref_pinned_mut_ref_lifetime<'a>(self: &std::pin::Pin<&'a mut Self>) {}
+
+    pub fn by_mut_ref_pinned_mut_ref(self: &mut std::pin::Pin<&mut Self>) {}
+
+    pub fn by_mut_ref_pinned_mut_ref_lifetime(self: &mut std::pin::Pin<&'_ mut Self>) {}
```
Comment (Frank-III) [inline test_crates/method_self_receivers/src/lib.rs:46]: I used https://grep.app/ to  find more examples, and it seems there are some usage of `Pin<&'_ mut Self>`, `Pin<&'a mut Self>` so I added them to the test crate

File: src/adapter/receiver.rs [hunk 7]
```diff
@@ -1,10 +1,11 @@
 use rustdoc_types::{GenericArgs, Type};
+use std::borrow::Cow;
```
Comment (obi1kenobi) [inline src/adapter/receiver.rs:1]: nit (and I'll eventually figure out how to get rustfmt to do this) but for readability it's usually a good idea to have built-in imports (`std / core` etc.) at the top then 3rd party imports afterward.

File: src/adapter/receiver.rs [hunk 8]
```diff
@@ -68,25 +69,36 @@ fn extract_kind_string(ty: &Type) -> String {
...
+                                }
+                                _ => unreachable!("should not encounter infer"),
                             })
                             .collect();
-                        format!("{}<{}>", name, args_str.join(", "))
+
+                        let args_joined = args_str
+                            .iter()
+                            .map(|cow| cow.as_ref())
+                            .collect::<Vec<_>>()
```
Comment (obi1kenobi) [inline src/adapter/receiver.rs:94]: Do we have to `.collect()` first to join, or can we join directly without allocating first?

File: src/adapter/properties.rs [hunk 9]
```diff
@@ -312,20 +312,20 @@ pub(super) fn resolve_method_receiver_property<'a, V: AsVertex<Vertex<'a>> + 'a>
...
         "by_mut_reference" => resolve_property_with(contexts, |vertex| {
-            let receiver = vertex.as_method_receiver().expect("not a MethodReceiver");
+            let receiver = vertex.as_receiver().expect("not a MethodReceiver");
             receiver.by_mut_reference().into()
         }),
         "kind" => resolve_property_with(contexts, |vertex| {
-            let receiver = vertex.as_method_receiver().expect("not a MethodReceiver");
-            receiver.kind().into()
+            let receiver = vertex.as_receiver().expect("not a MethodReceiver");
+            receiver.kind().as_ref().into()
```
Comment (obi1kenobi) [inline src/adapter/properties.rs]: Almost certainly already on your radar, but just in case: we want all the `.expect()` messages to reference `Receiver` instead of `MethodReceiver`.

**Review by obi1kenobi — COMMENTED** [review]
Excellent idea on the new test cases, and I love the initiative to proactively find more edge cases! 💯

A couple more nits and this is good to go.

File: src/adapter/receiver.rs (hunk 7 above)
Comment (Frank-III) [reply src/adapter/receiver.rs:1]: I see, good point!

File: src/adapter/receiver.rs (hunk 8 above)
Comment (Frank-III) [reply src/adapter/receiver.rs:94]: yes we need to `.collect()`

**Review by obi1kenobi — APPROVED** [review]
Excellent work!
